### PR TITLE
add TokenStreamRewriter to Python2 runtime

### PR DESCRIPTION
**Pull request by lygav**
Hello, i've added the TokenStreamRewriter to Python2 runtime.
If all goes well i hope to expand on this contribution.

Please advise about how to integrate into the auto-generated test suit if this is necessary.

P.S: Closes ISSUE #1115

### CODE REVIEW

**Comment by parrt** [comment]
@ericvergnaud this would be a nice addition. can you bless it?

**Comment by ericvergnaud** [comment]
blessed

**Comment by ericvergnaud** [comment]
blessed

File: runtime/Python2/src/antlr4/TokenStreamRewriter.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        for i, iop in enumerate(rewrites):
+            if any((iop is None, not isinstance(iop, TokenStreamRewriter.InsertBeforeOp))):
+                continue
+            prevInserts = [op for op in rewrites[:i] if isinstance(iop, TokenStreamRewriter.InsertBeforeOp)]
+            for prevIop in prevInserts:
+                if prevIop.index == iop.index:
+                    iop.text += prevIop.text
+                    rewrites[i] = None
+            # look for replaces where iop.index is in range; error
+            prevReplaces = [op for op in rewrites[:i] if isinstance(rop, TokenStreamRewriter.ReplaceOp)]
```
Comment (ericvergnaud) [inline runtime/Python2/src/antlr4/TokenStreamRewriter.py]: should this not be is instance(op, ...

**Comment by ericvergnaud** [comment]
@lygav please add some tests, following https://github.com/antlr/antlr4/blob/master/doc/adding-tests.md

**Comment by lygav** [comment]
@ericvergnaud Thank you for taking time to review the PR.
I'll add the needed tests.

File: runtime/Python2/src/antlr4/TokenStreamRewriter.py (hunk 1 above)
Comment (lygav) [inline runtime/Python2/src/antlr4/TokenStreamRewriter.py]: Seems you are right.
I'll fix that, and will add tests and will review the code again.

**Comment by lygav** [comment]
@ericvergnaud Hello, although TokenStreamRewriter belongs to the 'runtime' it looks like  TokenStreamRewriter has tests only as part of the "tool" testsuite, and only for Java.
The only other implementation is in C# 'runtime' and i just cannot see where is the test for it?
How do you propose to create test for it then:
 - create python test in the Pyhon2 runtime dir (will not run with the build?)
 - try to create TokenStreamDescriptor (or similar name..)

**Comment by ericvergnaud** [comment]
Hi,
create TokenStreamRewriterDescriptor is the way to go
Eric

**Comment by lygav** [comment]
Thanks Eric, i am going to try to figure this out.

**Comment by parrt** [comment]
@lygav I'll move my java tests to runtime.

**Comment by parrt** [comment]
Actually, it looks like a major pain to make this general. So far we test ANTLR itself and not runtime lib stuff that is accessed only with code. I'm not sure it's worth the effort for me at this time to make a general mechanism. It would be easier to have runtime tests access just library stuff in the target language. Better than not having tests I guess. @ericvergnaud open to suggestions.

**Comment by ericvergnaud** [comment]
Ok so let's leave as is and create independent tests in the respective runtimes.
We'll see on the long term whether there is a pattern that is worth sharing across runtimes.

**Comment by lygav** [comment]
Hello @ericvergnaud and @parrt , please look into the pull request with all the tests
replicated after the Java runtime.
Plus further enhancements to make Python2 runtime to be of the same consistent feel.
Also fixed another issue from some time ago (#550)

Thank you

**Pull request merged**